ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Offset.scala

```diff
@@ -3,47 +3,32 @@
 
 package com.daml.ledger.participant.state.v1
 
-import com.digitalasset.daml.lf.data.Ref.LedgerString
-import com.digitalasset.daml.lf.data.Ref
+import com.google.protobuf.ByteString
 
 /** Offsets into streams with hierarchical addressing.
   *
   * We use these [[Offset]]'s to address changes to the participant state.
-  * We allow for array of [[Long]] to allow for hierarchical addresses.
-  * These [[Long]] values are expected to be positive. Offsets are ordered by
-  * lexicographic ordering of the array elements.
+  * Offsets are opaque values that must be must be strictly
+  * increasing according to lexicographical ordering.
   *
-  * A typical use case for [[Offset]]s would be addressing a transaction in a
-  * blockchain by `[<blockheight>, <transactionId>]`. Depending on the
-  * structure of the underlying ledger these offsets are more or less
-  * nested, which is why we use an array of [[Long]]s. The expectation is
-  * though that there usually are few elements in the array.
+  * Ledger implementations are advised to future proof their design
+  * of offsets by reserving the first (few) bytes for a version
+  * indicator, followed by the specific offset scheme for that version.
+  * This way it is possible in the future to switch to a different versioning
+  * scheme, while making sure that previously created offsets are always
+  * less than newer offsets.
   *
   */
-final case class Offset(private val xs: Array[Long]) extends Ordered[Offset] {
-  def toLedgerString: Ref.LedgerString =
-    // It is safe to concatenate number and "-" to obtain a valid transactionId
-    Ref.LedgerString.assertFromString(components.mkString("-"))
+final class Offset(val value: ByteString) extends AnyVal with Ordered[Offset] {
 
-  override def toString: String = toLedgerString
+  override def compare(that: Offset): Int =
+    ByteString.unsignedLexicographicalComparator().compare(value, that.value)
```

Comment:
Very minor and likely to be inlined by the runtime, but probably worth calling `ByteString.unsignedLexicographicalComparator()` just once in a companion object.